feat(#1095): Update data sync documentation (#1243)
medic/cht-docs: content/en/apps/concepts/navigation.md

{{< figure src="sync-status.png" link="sync-status.png" class="right col-6 col-lg-4" >}}

Data synchronization is important for [offline users]({{< relref "apps/concepts/users" >}}). These users keep a copy of the data they have access to on their device. They can work from their device while disconnected from the internet (offline), by reading from and writing to their copy of the data. “Sync” (synchronization) is when data on the device is made to match the data on the server and requires an internet connection. The CHT app monitors the online status and attempts sync accordingly.

#### Replication Types

Synchronization consists of upward replication and downward replication. 
- **Upward Replication**: Uploading all new or updated data from the device to the server. It includes a retry mechanism for handling larger data batches, ensuring a robust and reliable upload process.
- **Downward Replication**: Downloading new or updated data from the server to the device. Downward replication may include the download of software updates to the CHT app when available.

> The CHT application manages data synchronization across two types of databases: the main application database and a meta database.

Reviewer: ```suggestion
The CHT application manages data synchronization across two types of databases:
```

I don't think this is necessary because we list them out next.

- **Main Application Database (`medic`)**: The main database that stores the primary data used by the application. It includes contacts, reports, messages, and other critical documents necessary for the core functionality of the application. This database is synchronized continuously to reflect changes to the application, such as new contact creations. Each user stores a subset of the main database which includes only the documents they're allowed to view.
- **User-Metadata Database(`medic-user-{username}-meta`)**: Each user has a dedicated database that stores operational metadata, including [telemetry data]({{< relref "apps/guides/performance/telemetry" >}}) and error messages. Synchronization occurs at predefined intervals to ensure up-to-date monitoring and analysis.

> The CHT application has two different Sync Intervals: 
> - **Regular Sync**: Occurs every 5 minutes. It replicates application data, including user-generated content and any changes to existing documents.
> - **Metadata Sync**: Occurs every 30 minutes, focusing on operational metadata rather than user-generated content. This ensures that telemetry data and system logs are regularly synchronized, providing insights into system performance and user activities

Reviewer: We should keep the names of the two databases consistent rather than using "regular" here. Suggestion:

```suggestion
The two databases sync on different schedules: the main application database syncs every 5 minutes, while the user metadata database syncs every 30 minutes.
```

Author: I deleted the content all-together because the sync interval is mentioned below in the Synchronization Triggers 

`- **Periodic Sync**: The application performs regular checks and attempts to synchronize. The main application database syncs every 5 minutes, while the user metadata database syncs every 30 minutes.`

It will also allow to retry the sync process in case of failure.

The user-initiated sync process initatied by clicking the "Sync now" buttoninvolves uploading local changes to the server and downloading updates from the server to the local database. It syncs both the main database and the metadata database to ensure that both user data and application metadata are up-to-date.

#### Synchronization Triggers
- **On Login**: Synchronization is automatically initiated upon successful user login if the app is connected to the internet. The CHT app checks for changes since the last known sequence, identifying document revisions that differ from the ones on the server, and sending or receiving bulk document updates.
- **Manual**: Clicking the "Sync now" button.
- **Periodic Sync**: The application performs regular checks and background processes that may trigger synchronization if there are local changes to replicate to the server, or changes from the server that need to be replicated to the local database.
- **On Reload**: Synchronization is automatically initiated when you click the reload button in the "Update available" modal.

> 

Reviewer: Also add one more...

```suggestion
- **On Connect**: The app also detects when an internet connection becomes available and attempts to sync immediately.
```

#### Sync Status States
The synchronization process can be in one of the following states:

- **Unknown**: The sync status is not determined yet.

> - **Disabled**: Sync is disabled, typically when the user is in online-only mode.

Reviewer: ```suggestion
- **Disabled**: Sync is disabled - only applies to online-only users.
```

- **InProgress**: Synchronization is currently ongoing.
- **Success**: The last sync operation was successful.
- **Required**: There is data pending synchronization.

#### Data Synchronization Retry Mechanism 

The CHT app has a retry mechanism to handle replication failures, particularly due to forbidden documents and challenging network environments with restricted data access. CHT core is configured to attempt replication retry up to a maximum of 3 times for each document.

> 

Author: I inferred it from the [db-sync-retry.service.ts](https://github.com/medic/cht-core/blob/d93b9162e21b180b1c920d58a6890812ae3ea9e7/webapp/src/ts/services/db-sync-retry.service.ts#L5C26-L5C26) file. 

Was my understanding incorrect ?

> 

Reviewer: Ahh I see! I think this is an implementation detail and not worth complicating these docs for. Essentially if the replication happens in the wrong order you might sync a report that the server thinks you shouldn't have access to. If that happens we change the order and try again.

> 

Author: Got it, in that case let me delete that section.
